kmip/services/server/engine.py

```diff
@@ -1509,22 +1509,37 @@ def _process_destroy(self, payload):
         else:
             unique_identifier = self._id_placeholder
 
-        self._get_object_with_access_controls(
+        managed_object = self._get_object_with_access_controls(
             unique_identifier,
             enums.Operation.DESTROY
         )
 
-        # TODO (peterhamilton) Process attributes to see if destroy possible
-        # 1. Check object state. If invalid, error out.
-        # 2. Check object deactivation date. If invalid, error out.
-
-        self._logger.info(
-            "Destroying an object with ID: {0}".format(unique_identifier)
-        )
+        # TODO If in "ACTIVE" state, we need to check its "Deactivation date"
+        # to see whether it can be destroyed or not
+        if hasattr(managed_object, 'state'):
+            if managed_object.state != enums.State.PRE_ACTIVE and \
+               managed_object.state != enums.State.DEACTIVATED and \
+               managed_object.state != enums.State.COMPROMISED:
+                raise exceptions.PermissionDenied(
+                    "Object is not pre-active, deactivated or compromised "
+                    "and cannot be destroyed."
+                )
 
-        self._data_session.query(objects.ManagedObject).filter(
-            objects.ManagedObject.unique_identifier == unique_identifier
-        ).delete()
+        # 'OpaqueObject' object has no attribute 'state'
+        if hasattr(managed_object, 'state') and \
+           managed_object.state == enums.State.COMPROMISED:
+            self._logger.info(
+                "Destroying a compromised object with ID: {0} (put into "
+                "Destroyed Compromised state)".format(unique_identifier)
+            )
+            managed_object.state = enums.State.DESTROYED_COMPROMISED
+        else:
+            self._logger.info(
+                "Destroying an object with ID: {0}".format(unique_identifier)
+            )
+            self._data_session.query(objects.ManagedObject).filter(
+                objects.ManagedObject.unique_identifier == unique_identifier
+            ).delete()
```

Comment:
Let's condense this a little bit. Let's remove the first log statement, so the `if` just sets `DESTROYED_COMPROMISED`. Then we can drop the `else:` line and keep the rest, letting the log and delete lines happen regardless of state. That way, `DESTROYED_COMPROMISED` objects still get deleted and the logs are consistent. This also makes it easier if we want to add special logic around when we actually delete objects down the road.